CMakeLists.txt

```diff
@@ -195,10 +195,91 @@ target_link_libraries(substitution_loader nlohmann_json::nlohmann_json)
 #message("FLEXFLOW_INCLUDE_DIRS: ${FLEXFLOW_INCLUDE_DIRS}")
 
 # compile flexflow lib
-if(BUILD_SHARED_LIBS)
-  cuda_add_library(flexflow SHARED ${FLEXFLOW_GPU_SRC} ${FLEXFLOW_SRC} OPTIONS ${CUDA_GENCODE})
+if (FF_GPU_BACKEND STREQUAL "cuda")
+  file(GLOB_RECURSE FLEXFLOW_GPU_SRC
+    LIST_DIRECTORIES False
+    ${FLEXFLOW_ROOT}/src/*.cu)
+
+  if(BUILD_SHARED_LIBS)
+    cuda_add_library(flexflow SHARED ${FLEXFLOW_GPU_SRC} ${FLEXFLOW_SRC} OPTIONS ${CUDA_GENCODE})
+  else()
+    cuda_add_library(flexflow STATIC ${FLEXFLOW_GPU_SRC} ${FLEXFLOW_SRC} OPTIONS ${CUDA_GENCODE})
+  endif()
+elseif(FF_GPU_BACKEND STREQUAL "hip_cuda" OR FF_GPU_BACKEND STREQUAL "hip_rocm")
+  file(GLOB_RECURSE FLEXFLOW_GPU_SRC
+    LIST_DIRECTORIES False
+    ${FLEXFLOW_ROOT}/src/*.cpp)
+
+  if(BUILD_SHARED_LIBS)
+    add_library(flexflow SHARED ${FLEXFLOW_GPU_SRC} ${FLEXFLOW_SRC})
+  else()
+    add_library(flexflow STATIC ${FLEXFLOW_GPU_SRC} ${FLEXFLOW_SRC})
+  endif()
+
+  list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/hip ${ROCM_PATH})
+
+  find_package(hip)
+
+  if (FF_GPU_BACKEND STREQUAL "hip_cuda")
+    # The targets defined by the hip cmake config only target amd devices.
+    # For targeting nvidia devices, we'll make our own interface target,
+    # hip_device_nvidia, that includes the rocm and hip headers. 
+    add_library(hip_device_nvidia INTERFACE)
+
+    # Configuring hipcc for nvidia:
+    #
+    # The platform hipcc targets is configured by the HIP_PLATFORM env var.
+    # Ideally, as we could in the shell, we would call `HIP_PLATFORM=nvidia hipcc <...>`.
+    # However, CMAKE_CXX_COMPILER doesn't allow configuration as such. Additionally,
+    # cmake doesn't allow setting environment variables for target builds like make does
+    # with exported variables.
+    #
+    # We write out a file, `nvidia_hipcc`, into the build directory to handle configuring hipcc. 
+    # nvidia_hipcc uses `cmake -E env` to agnostically set HIP_PLATFORM before calling hipcc.
+    cmake_path(APPEND _nvidia_hipcc "${CMAKE_CURRENT_BINARY_DIR}" nvidia_hipcc)
+    file(WRITE ${_nvidia_hipcc} "${CMAKE_COMMAND} -E env HIP_PLATFORM=${HIP_PLATFORM} ${HIP_HIPCC_EXECUTABLE} $@\n")
+    file(CHMOD ${_nvidia_hipcc} 
+      PERMISSIONS 
+        OWNER_READ OWNER_WRITE OWNER_EXECUTE 
+        GROUP_READ GROUP_EXECUTE 
+        WORLD_READ WORLD_EXECUTE)
```

Comment:
@lockshaw This is a bit hacky. Alternatively we commit a script into source that does the same thing. Do you have a preference?

Reply:
I'd probably lean toward the cmake approach, but I don't have a particularly strong feeling either way. Is there any difference in terms of impact on developer experience between the two?

Reply:
I don't think there's a difference in developer experience, both are grockable. +1 to cmake approach, will leave as is